File: internal/service/ticket_service.go
package service

import (
	"github.com/google/uuid"

	"github.com/tiketin/backend/internal/dto"
	"github.com/tiketin/backend/internal/model"
	"github.com/tiketin/backend/internal/repository"
)

type TicketService struct {
	ticketRepo *repository.TicketRepository
}

func NewTicketService(ticketRepo *repository.TicketRepository) *TicketService {
	return &TicketService{ticketRepo: ticketRepo}
}

func (s *TicketService) Purchase(userID string, req dto.PurchaseTicketRequest) (*model.Ticket, error) {
	ticket := &model.Ticket{
		ID:      uuid.NewString(),
		EventID: req.EventID,
		UserID:  userID,
		Status:  model.TicketStatusPending,
	}

	if err := s.ticketRepo.Create(ticket); err != nil {
		return nil, err
	}
	return ticket, nil
}

func (s *TicketService) GetMyTickets(userID string) ([]model.Ticket, error) {
	return s.ticketRepo.FindByUserID(userID)
}

File: src/services/ticket.service.ts
import { TicketRepository } from "@/src/repositories/ticket.repository";
import { EventRepository } from "@/src/repositories/event.repository";
import { PurchaseTicketRequest } from "@/src/validators/ticket.validator";

export class TicketService {
  static async getMyTickets(userId: string) {
    return TicketRepository.findByUserId(userId);
  }

  static async purchase(userId: string, req: PurchaseTicketRequest) {
    const { event_id, quantity } = req;

    const quota = await EventRepository.findQuotaById(event_id);
    if (quota === null) {
      throw new Error("Event not found");
    }
    if (quota < quantity) {
      throw new Error("Not enough quota");
    }

    const tickets = [];
    for (let i = 0; i < quantity; i++) {
      const ticket = await TicketRepository.create({
        eventId: event_id,
        userId,
        status: "paid",
      });
      tickets.push(ticket);
    }

    await EventRepository.decrementQuota(event_id, quota - quantity);

    return { ticket_ids: tickets.map((t) => t.id) };
  }
}
